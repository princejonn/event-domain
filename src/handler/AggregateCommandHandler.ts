import Joi from "joi";
import {
  IAggregateCommandHandler,
  AggregateCommandHandlerContext,
  AggregateCommandHandlerOptions,
  HandlerConditions,
  HandlerIdentifier,
} from "../types";

export class AggregateCommandHandler<State> implements IAggregateCommandHandler<State> {
  readonly aggregate: HandlerIdentifier;
  readonly commandName: string;
  readonly conditions: HandlerConditions;
  readonly schema: Joi.Schema;
  readonly handler: (context: AggregateCommandHandlerContext<State>) => Promise<void>;

  constructor(options: AggregateCommandHandlerOptions<State>) {
    this.aggregate = { name: options.aggregate?.name, context: options.aggregate?.context };
    this.commandName = options.commandName || null;
    this.conditions = options.conditions || {};
    this.schema = options.schema;
    this.handler = options.handler;
  }
}
