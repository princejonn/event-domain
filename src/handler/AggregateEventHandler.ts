import {
  IAggregateEventHandler,
  AggregateEventHandlerContext,
  AggregateEventHandlerOptions,
  HandlerIdentifier,
} from "../types";

export class AggregateEventHandler<State> implements IAggregateEventHandler<State> {
  readonly aggregate: HandlerIdentifier;
  readonly eventName: string;
  readonly handler: (context: AggregateEventHandlerContext<State>) => Promise<void>;

  constructor(options: AggregateEventHandlerOptions<State>) {
    this.aggregate = { name: options.aggregate?.name, context: options.aggregate?.context };
    this.eventName = options.eventName || null;
    this.handler = options.handler;
  }
}
