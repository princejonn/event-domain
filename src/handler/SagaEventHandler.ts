import {
  HandlerConditions,
  HandlerIdentifier,
  HandlerIdentifierMultipleContexts,
  ISagaEventHandler,
  SagaEventHandlerContext,
  SagaEventHandlerOptions,
  SagaStoreSaveOptions,
  GetSagaIdFunction,
} from "../types";

export class SagaEventHandler<State> implements ISagaEventHandler<State> {
  readonly aggregate: HandlerIdentifierMultipleContexts;
  readonly conditions: HandlerConditions;
  readonly eventName: string;
  readonly getSagaId: GetSagaIdFunction;
  readonly saga: HandlerIdentifier;
  readonly saveOptions: SagaStoreSaveOptions;
  readonly handler: (context: SagaEventHandlerContext<State>) => Promise<void>;

  constructor(options: SagaEventHandlerOptions<State>) {
    this.aggregate = { name: options.aggregate.name, context: options.aggregate.context };
    this.conditions = options.conditions || {};
    this.eventName = options.eventName;
    this.getSagaId = options.getSagaId;
    this.saga = { name: options.saga?.name, context: options.saga?.context };
    this.saveOptions = options.saveOptions || {};
    this.handler = options.handler;
  }
}
