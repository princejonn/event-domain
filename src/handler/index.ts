export * from "./AggregateCommandHandler";
export * from "./AggregateEventHandler";
export * from "./SagaEventHandler";
export * from "./ViewEventHandler";
