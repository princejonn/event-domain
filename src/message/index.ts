export * from "./Command";
export * from "./DomainEvent";
export * from "./Message";
export * from "./Timeout";
