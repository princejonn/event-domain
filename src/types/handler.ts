export interface HandlerIdentifier {
  name: string;
  context: string;
}

export interface HandlerConditions {
  created?: boolean;
  permanent?: boolean;
}

export interface HandlerIdentifierExpectingStructure {
  name?: string;
  context?: string;
}

export interface HandlerIdentifierMultipleContexts {
  name: string;
  context?: Array<string> | string;
}
