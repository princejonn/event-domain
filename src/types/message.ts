import { AggregateIdentifier } from "./aggregate";
import { MessageType } from "../enum";

export interface IMessage {
  aggregate: AggregateIdentifier;
  causationId: string;
  correlationId: string;
  data: Record<string, any>;
  delay: number;
  id: string;
  mandatory: boolean;
  name: string;
  timestamp: Date;
  type: MessageType;
}

export interface MessageOptions {
  aggregate: AggregateIdentifier;
  causationId?: string;
  correlationId?: string;
  data?: Record<string, any>;
  delay?: number;
  id?: string;
  mandatory?: boolean;
  name: string;
  timestamp?: Date | string;
  type?: MessageType;
}
