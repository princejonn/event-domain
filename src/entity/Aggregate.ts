import { AGGREGATE_MERGE_STATE_SCHEMA, getApplyEventSchema } from "../schema";
import { AggregateEventHandler } from "../handler";
import { Command, DomainEvent } from "../message";
import { Logger } from "@lindorm-io/winston";
import { assertCamelCase, validateSchemaAsync, validateSchemaSync } from "../util";
import { cloneDeep, find, merge, set } from "lodash";
import {
  AggregateData,
  AggregateEventHandlerContext,
  AggregateIdentifier,
  AggregateOptions,
  AnyType,
  GenericRecord,
  IAggregate,
} from "../types";
import {
  AggregateDestroyedError,
  AggregateNotDestroyedError,
  HandlerNotRegisteredError,
  IllegalEntityChangeError,
  MessageTypeError,
} from "../error";

export class Aggregate<State> implements IAggregate<State> {
  readonly id: string;
  readonly name: string;
  readonly context: string;

  private readonly _eventHandlers: Array<AggregateEventHandler<State>>;
  private readonly _events: Array<DomainEvent>;
  private readonly _state: GenericRecord<State>;
  private _destroyed: boolean;
  private _destroying: boolean;
  private _numberOfLoadedEvents: number;

  private readonly logger: Logger;

  constructor(options: AggregateOptions<State>, logger: Logger) {
    this.logger = logger.createChildLogger(["Aggregate"]);

    assertCamelCase(options.name);

    this.id = options.id;
    this.name = options.name;
    this.context = options.context;

    this._destroyed = false;
    this._destroying = false;
    this._eventHandlers = options.eventHandlers || [];
    this._events = [];
    this._numberOfLoadedEvents = 0;
    this._state = {};
  }

  public get destroyed(): boolean {
    return this._destroyed;
  }
  public set destroyed(_) {
    throw new IllegalEntityChangeError();
  }

  public get events(): Array<DomainEvent> {
    return this._events;
  }
  public set events(_) {
    throw new IllegalEntityChangeError();
  }

  public get numberOfLoadedEvents(): number {
    return this._numberOfLoadedEvents;
  }
  public set numberOfLoadedEvents(_) {
    throw new IllegalEntityChangeError();
  }

  public get state(): GenericRecord<State> {
    return this._state;
  }
  public set state(_) {
    throw new IllegalEntityChangeError();
  }

  public async apply(
    causation: Command,
    name: string,
    data?: Record<string, any>,
  ): Promise<void> {
    this.logger.debug("applying command", { causation, name, data });

    await this.handleEvent(
      new DomainEvent(
        {
          aggregate: { id: this.id, name: this.name, context: this.context },
          data: data,
          name: name,
        },
        causation,
      ),
    );
  }

  public async load(event: DomainEvent): Promise<void> {
    this.logger.debug("loading event", { event });

    await this.handleEvent(event);
    this._numberOfLoadedEvents += 1;
  }

  private async handleEvent(event: DomainEvent): Promise<void> {
    this.logger.debug("handling event", { event });

    if (!(event instanceof DomainEvent)) {
      throw new MessageTypeError(event, DomainEvent);
    }

    if (this._destroyed) {
      throw new AggregateDestroyedError();
    }

    const destroying = this._destroying;

    await validateSchemaAsync(
      getApplyEventSchema({
        id: this.id,
        name: this.name,
        context: this.context,
      }),
      {
        name: event.name,
        aggregate: event.aggregate,
        data: event.data,
        causationId: event.causationId,
      },
    );

    const eventHandler: AggregateEventHandler<State> = find(
      this._eventHandlers,
      (handler) => handler.eventName === event.name,
    );

    if (!(eventHandler instanceof AggregateEventHandler)) {
      throw new HandlerNotRegisteredError();
    }

    const context: AggregateEventHandlerContext<State> = {
      event: event,
      state: cloneDeep(this._state),
      logger: this.logger.createChildLogger(["AggregateEventHandler"]),
      destroy: this.destroy.bind(this),
      destroyNext: this.destroyNext.bind(this),
      mergeState: this.mergeState.bind(this),
      setState: this.setState.bind(this),
    };

    await eventHandler.handler(context);

    if (destroying && !this._destroyed) {
      throw new AggregateNotDestroyedError();
    }

    this._events.push(event);

    this.logger.debug("successfully handled event", { event });
  }

  private destroy(): void {
    if (this._destroyed) {
      throw new AggregateDestroyedError();
    }

    this._destroyed = true;
    this._destroying = false;
  }

  private destroyNext(): void {
    if (this._destroyed) {
      throw new AggregateDestroyedError();
    }

    this._destroying = true;
  }

  private mergeState(data: Record<string, any>): void {
    validateSchemaSync(AGGREGATE_MERGE_STATE_SCHEMA, { data });

    if (this._destroyed) {
      throw new AggregateDestroyedError();
    }

    merge(this._state, data);
  }

  private setState(path: string, value: AnyType): void {
    if (this._destroyed) {
      throw new AggregateDestroyedError();
    }

    set(this._state, path, value);
  }

  public toJSON(): AggregateData<State> {
    return {
      id: this.id,
      name: this.name,
      context: this.context,
      destroyed: this.destroyed,
      events: cloneDeep(this.events),
      numberOfLoadedEvents: cloneDeep(this.numberOfLoadedEvents),
      state: cloneDeep(this.state),
    };
  }

  static identifier(identifier: AggregateIdentifier): string {
    return `${identifier.context}.${identifier.name}.${identifier.id}`;
  }
}
