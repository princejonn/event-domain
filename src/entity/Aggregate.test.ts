import MockDate from "mockdate";
import { Aggregate } from "./Aggregate";
import { AggregateEventHandler } from "../handler";
import { Command, DomainEvent } from "../message";
import {
  AggregateDestroyedError,
  AggregateNotDestroyedError,
  IllegalEntityChangeError,
} from "../error";
import {
  aggregateIdentifier,
  commandWithData,
  commandWithDestroy,
  commandWithDestroyNext,
  eventHandlerWithData,
  eventHandlerWithDestroy,
  eventHandlerWithDestroyNext,
  logger,
} from "../test";

MockDate.set("2022-01-01T08:00:00.000Z");

jest.mock("crypto", () => ({
  randomUUID: jest.fn(() => "e64b576a-022e-4dd4-8c97-32408012f8d1"),
}));

describe("Aggregate", () => {
  describe("when creating an aggregate with camelCase name", () => {
    it("should create aggregate", () => {
      expect(
        () =>
          new Aggregate(
            {
              id: "id",
              name: "camelCasedName",
              context: "default",
              eventHandlers: [],
            },
            logger,
          ),
      ).not.toThrow();
    });
  });

  describe("when creating an aggregate with non-camelCase name", () => {
    it("should throw", () => {
      expect(
        () =>
          new Aggregate(
            {
              id: "id",
              name: "snake_case_name",
              context: "default",
              eventHandlers: [],
            },
            logger,
          ),
      ).toThrow(new Error("snake_case_name is not in camelCase"));
    });
  });

  describe("when applying a command with data", () => {
    let aggregate: Aggregate<any>;

    beforeAll(async () => {
      aggregate = new Aggregate(
        {
          ...aggregateIdentifier,
          eventHandlers: [
            new AggregateEventHandler({
              eventName: "eventWithData",
              aggregate: aggregateIdentifier,
              handler: async (ctx) => {
                ctx.mergeState(ctx.event.data);
                ctx.setState("OBJ.STR", "value");
                ctx.setState("OBJ.OBJ.NUM", 123456);
              },
            }),
          ],
        },
        logger,
      );

      // @ts-ignore
      await aggregate.setState("OBJ.STR", "overwrite");
      // @ts-ignore
      await aggregate.setState("name", { bar: "baz" });

      await aggregate.apply(
        new Command({
          id: "6c2e5bbd-6643-469d-92ba-f6a140d42cb4",
          name: "commandWithData",
          data: { name: "hello" },
          aggregate: aggregateIdentifier,
        }),
        "eventWithData",
        { name: { first: "hello", foo: "bar" } },
      );
    });

    it("should add events to array", () => {
      expect(aggregate.events).toMatchSnapshot();
    });

    it("should set state and overwrite old values", () => {
      expect(aggregate.state.OBJ.STR).toBe("value");
    });

    it("should set state with new values", () => {
      expect(aggregate.state.OBJ.OBJ.NUM).toBe(123456);
    });

    it("should merge state with new values", () => {
      expect(aggregate.state.name.first).toBe("hello");
      expect(aggregate.state.name.foo).toBe("bar");
    });

    it("should merge state with old values", () => {
      expect(aggregate.state.name.bar).toBe("baz");
    });
  });

  describe("when applying a command with destroyNext", () => {
    let aggregate: Aggregate<any>;

    beforeAll(async () => {
      aggregate = new Aggregate(
        {
          ...aggregateIdentifier,
          eventHandlers: [
            eventHandlerWithData,
            eventHandlerWithDestroy,
            eventHandlerWithDestroyNext,
          ],
        },
        logger,
      );

      await aggregate.apply(commandWithDestroyNext, "eventWithDestroyNext", {});
    });

    it("should expect aggregate to be destroyed next", () => {
      // @ts-ignore
      expect(aggregate._destroying).toBe(true);
    });

    it("should expect aggregate to not be destroyed yet", () => {
      expect(aggregate.destroyed).toBe(false);
    });
  });

  describe("when applying a command with destroy", () => {
    let aggregate: Aggregate<any>;

    beforeAll(async () => {
      aggregate = new Aggregate(
        {
          ...aggregateIdentifier,
          eventHandlers: [
            eventHandlerWithData,
            eventHandlerWithDestroy,
            eventHandlerWithDestroyNext,
          ],
        },
        logger,
      );

      await aggregate.apply(commandWithDestroy, "eventWithDestroy", {});
    });

    it("should set aggregate as destroyed", () => {
      expect(aggregate.destroyed).toBe(true);
    });
  });

  describe("when applying a command with destroy after destroyNext", () => {
    let aggregate: Aggregate<any>;

    beforeAll(async () => {
      aggregate = new Aggregate(
        {
          ...aggregateIdentifier,
          eventHandlers: [
            eventHandlerWithData,
            eventHandlerWithDestroy,
            eventHandlerWithDestroyNext,
          ],
        },
        logger,
      );

      await aggregate.apply(commandWithDestroyNext, "eventWithDestroyNext", {});
      await aggregate.apply(commandWithDestroy, "eventWithDestroy", {});
    });

    it("should set destroying to false", () => {
      // @ts-ignore
      expect(aggregate._destroying).toBe(false);
    });

    it("should expect aggregate to be destroyed", () => {
      expect(aggregate.destroyed).toBe(true);
    });
  });

  describe("when applying a command with data after destroyNext", () => {
    let aggregate: Aggregate<any>;

    beforeAll(async () => {
      aggregate = new Aggregate(
        {
          ...aggregateIdentifier,
          eventHandlers: [
            eventHandlerWithData,
            eventHandlerWithDestroy,
            eventHandlerWithDestroyNext,
          ],
        },
        logger,
      );

      await aggregate.apply(commandWithDestroyNext, "eventWithDestroyNext", {});
    });

    it("should throw AggregateNotDestroyedError", async () => {
      await expect(
        aggregate.apply(commandWithData, "eventWithData", commandWithData.data),
      ).rejects.toThrow(AggregateNotDestroyedError);
    });
  });

  describe("when applying a command with data after destroy", () => {
    let aggregate: Aggregate<any>;

    beforeAll(async () => {
      aggregate = new Aggregate(
        {
          ...aggregateIdentifier,
          eventHandlers: [
            eventHandlerWithData,
            eventHandlerWithDestroy,
            eventHandlerWithDestroyNext,
          ],
        },
        logger,
      );

      await aggregate.apply(commandWithDestroy, "eventWithDestroy", {});
    });

    it("should throw AggregateDestroyedError", async () => {
      await expect(
        aggregate.apply(commandWithData, "eventWithData", commandWithData.data),
      ).rejects.toThrow(AggregateDestroyedError);
    });
  });

  describe("when loading an existing event", () => {
    let aggregate: Aggregate<any>;

    beforeAll(async () => {
      aggregate = new Aggregate(
        {
          ...aggregateIdentifier,
          eventHandlers: [
            eventHandlerWithData,
            eventHandlerWithDestroy,
            eventHandlerWithDestroyNext,
          ],
        },
        logger,
      );

      await aggregate.load(
        new DomainEvent({
          aggregate: aggregateIdentifier,
          causationId: "ce03d1ac-2f77-4941-88ce-2960aabf6400",
          correlationId: "2038c9e9-349e-4ac4-a750-cd474240c37e",
          data: { freddie: "mercury" },
          delay: 0,
          id: "aed46332-a369-48b7-a100-1a6bbdd13552",
          mandatory: true,
          name: "eventWithData",
          timestamp: new Date("2001-02-03T04:05:06.789Z"),
        }),
      );
    });

    it("should add event list of loaded events", () => {
      expect(aggregate.events).toMatchSnapshot();
    });

    it("should increment number of loaded events", () => {
      expect(aggregate.numberOfLoadedEvents).toBe(1);
    });

    it("should handle event and set state", () => {
      expect(aggregate.state).toMatchSnapshot();
    });
  });

  describe("when trying to set destroyed", () => {
    let aggregate: Aggregate<any>;

    beforeAll(() => {
      aggregate = new Aggregate({ ...aggregateIdentifier, eventHandlers: [] }, logger);
    });

    it("should throw", () => {
      expect(() => {
        aggregate.destroyed = true;
      }).toThrow(IllegalEntityChangeError);
    });
  });

  describe("when trying to set events", () => {
    let aggregate: Aggregate<any>;

    beforeAll(() => {
      aggregate = new Aggregate({ ...aggregateIdentifier, eventHandlers: [] }, logger);
    });

    it("should throw", () => {
      expect(() => {
        aggregate.events = [];
      }).toThrow(IllegalEntityChangeError);
    });
  });

  describe("when trying to set numberOfLoadedEvents", () => {
    let aggregate: Aggregate<any>;

    beforeAll(() => {
      aggregate = new Aggregate({ ...aggregateIdentifier, eventHandlers: [] }, logger);
    });

    it("should throw", () => {
      expect(() => {
        aggregate.numberOfLoadedEvents = 2;
      }).toThrow(IllegalEntityChangeError);
    });
  });

  describe("when trying to set state", () => {
    let aggregate: Aggregate<any>;

    beforeAll(() => {
      aggregate = new Aggregate({ ...aggregateIdentifier, eventHandlers: [] }, logger);
    });

    it("should throw", () => {
      expect(() => {
        aggregate.state = {};
      }).toThrow(IllegalEntityChangeError);
    });
  });
});
