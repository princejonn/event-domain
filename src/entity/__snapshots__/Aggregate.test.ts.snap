// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Aggregate when applying a command with data should add events to array 1`] = `
Array [
  DomainEvent {
    "aggregate": Object {
      "context": "defaultContext",
      "id": "2a553a74-0f84-4a7d-b91e-2902666db98c",
      "name": "aggregateName",
    },
    "causationId": "6c2e5bbd-6643-469d-92ba-f6a140d42cb4",
    "correlationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
    "data": Object {
      "name": Object {
        "first": "hello",
        "foo": "bar",
      },
    },
    "delay": 0,
    "id": "e64b576a-022e-4dd4-8c97-32408012f8d1",
    "mandatory": false,
    "name": "eventWithData",
    "timestamp": 2022-01-01T08:00:00.000Z,
    "type": "event",
  },
]
`;

exports[`Aggregate when loading an existing event should add event list of loaded events 1`] = `
Array [
  DomainEvent {
    "aggregate": Object {
      "context": "defaultContext",
      "id": "2a553a74-0f84-4a7d-b91e-2902666db98c",
      "name": "aggregateName",
    },
    "causationId": "ce03d1ac-2f77-4941-88ce-2960aabf6400",
    "correlationId": "2038c9e9-349e-4ac4-a750-cd474240c37e",
    "data": Object {
      "freddie": "mercury",
    },
    "delay": 0,
    "id": "aed46332-a369-48b7-a100-1a6bbdd13552",
    "mandatory": true,
    "name": "eventWithData",
    "timestamp": 2001-02-03T04:05:06.789Z,
    "type": "event",
  },
]
`;

exports[`Aggregate when loading an existing event should handle event and set state 1`] = `
Object {
  "OBJ": Object {
    "OBJ": Object {
      "NUM": 123456,
    },
    "STR": "value",
  },
  "freddie": "mercury",
}
`;
